feat: Parse tomes as gears, show tome identification percentages

kristofbolyai opened the pull request:
This PR is best reviewed per commit, to understand the underlying refactors needed for it to work. Some commits have descriptions, to help understanding the commits. The size of the PR is not that huge as it seems, especially when going through commits 1 by 1.

Tomes are really close to actual gear items, but they need some adaptations, hence the refactors. In the end, the new system behaves in a much more adaptive way, allowing us to implement new item types effectively.

Some screenshots:
![image](https://github.com/Wynntils/Artemis/assets/49001742/3c16661a-6ee2-48a1-a524-6e0dac1ec613)
![image](https://github.com/Wynntils/Artemis/assets/49001742/04cf1e68-8c62-42b5-a294-3473dd2ab6a3)
![image](https://github.com/Wynntils/Artemis/assets/49001742/0ee152a1-7552-429f-8b06-52836c5f52bb)
![image](https://github.com/Wynntils/Artemis/assets/49001742/400973be-788a-4534-b7cb-e117ac3b2c58)
![image](https://github.com/Wynntils/Artemis/assets/49001742/5293cf51-a15a-4a12-8fe6-ec2eb0d0a805)
![image](https://github.com/Wynntils/Artemis/assets/49001742/dc32595d-1ed2-4bc6-9645-06fde41bd8d6)

kristofbolyai committed:
refactor: Create AbstractItemInfoDeserializer, as a common API-info deserializer

magicus commented on common/src/main/java/com/wynntils/models/gear/GearInfoRegistry.java:
```diff
@@ -89,7 +68,7 @@ private void loadGearRegistry() {
         dl.handleJsonObject(json -> {
             Gson gson = new GsonBuilder()
-                    .registerTypeHierarchyAdapter(GearInfo.class, new GearInfoDeserializer())
+                    .registerTypeHierarchyAdapter(GearInfo.class, new GearInfoDeserizalier())
```
typo in the name

kristofbolyai committed:
feat: Load tome and parse tome API data, apply it to items

kristofbolyai committed:
refactor: WynnItemParser#parseItemStack only needs to know about stat possible values

kristofbolyai committed:
refactor: Rename TomeInfo to DeprecatedTomeInfo

kristofbolyai committed:
chore: Remove DeprecatedTomeInfo

kristofbolyai committed:
refactor: Promote (Gear)TooltipModel into an upper-level package

kristofbolyai committed:
refactor: Rename every related class to be "generic"

kristofbolyai committed:
refactor: Make TooltipModel and all of its components generic

This refactor prepares the tooltip model to add custom tooltips to other types of items that are identifiable, like tomes. If a new type of identifiable, non-gear like item is added, like tomes and charms, the model can be adapted to support the new item type, simply by implementing an `IdentifiableItemInfo` for the type.

kristofbolyai committed:
refactor: Extract StatCalculator#calculateOverallQuality, create TomeInstance

kristofbolyai committed:
feat: Make ItemStatInfoFeature work with tomes

magicus commented on common/src/main/java/com/wynntils/models/gear/type/AbstractItemInfoDeserializer.java:
```diff
@@ -0,0 +1,416 @@
+ * Copyright © Wynntils 2023.
+ * This file is released under LGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.models.gear.type;
```
Not sure the type package is the right place for this. We usually reserve it for interfaces, enums and records.

magicus replied:
I think it better belongs in `models.wynnitem`, an "abstract" model I tried to create to unify similar item kinds. I think this has the same kind of vibe to it.

magicus commented on common/src/main/java/com/wynntils/models/gear/type/AbstractItemInfoDeserializer.java:
```diff
@@ -0,0 +1,416 @@
+        String lore = JsonUtils.getNullableJsonString(json, "lore");
+        if (lore == null) return Optional.empty();
+
+        // Some lore contain like "\\[Community Event Winner\\]", fix that
```
Just noticed this: Is this still relevant in the v3 api?

kristofbolyai replied:
Removed it, it's not.

magicus commented on common/src/main/java/com/wynntils/models/tooltip/IdentifiableItemInfo.java:
```diff
@@ -0,0 +1,34 @@
+ * Copyright © Wynntils 2023.
+ * This file is released under LGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.models.tooltip;
```
move to `type` subpackage.

magicus commented on common/src/main/java/com/wynntils/models/tooltip/TooltipIdentificationDecorator.java:
```diff
@@ -2,14 +2,13 @@
  * This file is released under LGPLv3. See LICENSE for full license details.
  */
-package com.wynntils.models.gear.tooltip;
+package com.wynntils.models.tooltip;
```
Should also be in `type`, I think.

magicus commented on common/src/main/java/com/wynntils/models/tooltip/TooltipStyle.java:
```diff
@@ -2,11 +2,11 @@
  * This file is released under LGPLv3. See LICENSE for full license details.
  */
-package com.wynntils.models.gear.tooltip;
+package com.wynntils.models.tooltip;
```
`type`

magicus commented:
Call me crazy, but I'm seriously thinking that the new `tooltip` model would be better of as a `handler` than a model. According to the javadoc guidelines:

```
 * Models are our representation of the Wynncraft world, game state and Wynncraft
 * metadata. In general, any non-trivial interesting aspect of Wynncraft that is
 * provided through vanilla Minecraft elements, should be parsed and provided by
 * a Model.
```

and

```
 * Handlers span the bridge between Minecraft and Wynncraft. They manage a certain
 * aspect of Minecraft functionality, and with some Wynncraft knowledge, they distribute
 * the MC content to the actual Wynncraft models. A handler is needed when there is not a
 * clear 1-to-1 relationship between Minecraft components and models.
```

In this case it is not really "dispatching", but I think it is the same core idea. We need to provide a service to several models to be able to build tooltips. That is not so much a model per se, as a support layer for the models.

magicus commented:
Overall, I think this look pretty nifty!

kristofbolyai committed:
chore: Move around all classes to correct packages, from reviews

TooltipModel -> TooltipHandler
Tooltip.type package
AbstractItemInfoDeserializer -> WynnItem

kristofbolyai committed:
fix: Remove AbstractItemInfoDeserializer lore special handling for old API

kristofbolyai commented:
I also moved the TooltipModel to be a TooltipHandler. IMO it feels just as weird, but I can see how this is theoretically a Handler, it's just that we haven't made any of this kind.
